bin/npg_publish_tree.pl

```diff
@@ -188,46 +188,32 @@ sub handler {
 my @files = grep { -f } $publisher->list_directory($source_directory,
                                                    recurse => 1);
 
-my $mlwh_json_cb = sub {
-  # DataObject, path to file folder, collection file
-  my ($obj, $collection, $file) = @_;
-  if (defined $mlwh_json_filename) { 
+sub write_json {
+  my ($json_filename, $irods_collection) = @_;
+  if (defined $json_filename && (not $json_filename =~ qr/^\s*$/sxm)
+      && defined $irods_collection && (not $irods_collection =~ qr/^\s*$/sxm)) {
     Readonly::Scalar my $JSON_FILE_VERSION => '1.0';
-    my $mlwh_hash = {
-      irods_root_collection    => $collection,
-      irods_data_relative_path => $file
-    };
-    my ($json_fh, $json_hash);
-    if (-e $mlwh_json_filename) {
-      open $json_fh, '+<:encoding(UTF-8)', $mlwh_json_filename or
-        self->logcroak(q[could not open ml warehouse json file] .
-        qq[$mlwh_json_filename]);
-      $json_hash = decode_json <$json_fh>;
-    }
-    else {
-      open $json_fh, '>:encoding(UTF-8)', $mlwh_json_filename or
-        self->logcroak(q[could not open ml warehouse json file] .
-        qq[$mlwh_json_filename]);
-      $json_hash = {
-        version  => $JSON_FILE_VERSION,
-        irods_top_collection_folder => $irods_tmp_coll,
-        products => [],
-      };
-    }
-    push @{$json_hash->{products}}, $mlwh_hash;
-
-    seek $json_fh, 0, 0;
 
+    my ($json_fh, $json_hash);
+    open $json_fh, '>:encoding(UTF-8)', $json_filename or
+      self->logcroak(q[could not open ml warehouse json file] .
+      qq[$json_filename]);
+    $json_hash = {
+      version  => $JSON_FILE_VERSION,
+      irods_collection => $irods_collection
+    };
     print $json_fh encode_json($json_hash) or
       self->logcroak(q[could not write to ml warehouse json file ] .
-      qq[$mlwh_json_filename]);
+      qq[$json_filename]);
 
     close $json_fh or
       self->logcroak(q[could not close ml warehouse json file] .
-      qq[$mlwh_json_filename]);
+      qq[$json_filename]);
+  } else {
+    self->logcroak(q[Wrong parameters in write_json]);
```

Comment:
Error messages should say precisely what is wrong, if possible. Especially when they are feedback to users. See previous comment about where to validate these arguments.

It is better to use the form:

```
if (<test>) {
  <raise error>
}

<do some things>

```
than

```
if (<test>) {
    <do some things>
} else {
    <raise error>
}
```
because it eliminates the error case first and you don't have to refer back when reaching the `else` clause. It also saves a level of indentation.

Reply:
Ok thank you. I think we do not need that error anymore since the parameter checks have to be removed.